Local app config for Linux

cyphase opened the pull request:
<!--- Describe your changes in detail -->

Refactored tests to respect XDG app path customization, and added the functionality for the GTK backend while marking it as not implemented on the other backends.

<!--- What problem does this change solve? -->

This fixes the problem of Linux users not having their app files stored in the place they've configured.

<!--- If this PR relates to an issue, include Refs #XXX or Fixes #XXX -->

Fixes (partially) #2867 (since that mentions Windows as well).

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

freakboy3742 commented on gtk/tests_backend/app.py:
```diff
@@ -25,25 +25,42 @@ def __init__(self, app):
     def is_cursor_visible(self):
         pytest.skip("Cursor visibility not implemented on GTK")
 
+    def apply_path_customization(self):
+        root = Path.home() / ".cache/testbed/custom_app_paths"
+        os.environ["XDG_CONFIG_HOME"] = str(root / "config")
+        os.environ["XDG_DATA_HOME"] = str(root / "data")
+        os.environ["XDG_CACHE_HOME"] = str(root / "cache")
+        os.environ["XDG_STATE_HOME"] = str(root / "state")
+
+    def remove_path_customization(self):
+        del os.environ["XDG_CONFIG_HOME"]
+        del os.environ["XDG_DATA_HOME"]
+        del os.environ["XDG_CACHE_HOME"]
+        del os.environ["XDG_STATE_HOME"]
```
This might be better as a `contextlib.contextmanager`, so it can't be accidentally mis-used.

cyphase replied:
Done.

freakboy3742 reviewed (changes requested):
This needs a changenote, and there's one suggestion inline about using a context manager; but otherwise this looks great!

cyphase committed:
Refactor path tests to use a context manager.

This is in preparation for a future change.

cyphase committed:
Add customized app path testing to testbed.

No backends implement it yet.

cyphase committed:
Make the GTK backend respect XDG app path environment variables.

This commit also updates related documentation.

cyphase commented:
Does this need any documentation beyond the changenote?

rmartin16 commented:
I'm sure this is on Russell's radar...but as a doublecheck: while I agree this is a bugfix, I would also consider it a candidate for the `Backward Incompatible Changes` changelog section.

cyphase commented:
> I'm sure this is on Russell's radar...but as a doublecheck: while I agree this is a bugfix, I would also consider it a candidate for the `Backward Incompatible Changes` changelog section.

freakboy3742 commented:
> Does this need any documentation beyond the changenote?

I'd say it's worth adding a platform note to the [app paths docs](https://toga.readthedocs.io/en/stable/reference/api/resources/app_paths.html). That page doesn't have a notes section at present, but a lot of other widgets have a Notes section to capture "platform quirks" and things like that (see [MapView](https://toga.readthedocs.io/en/stable/reference/api/widgets/mapview.html#notes) for one example).

As @rmartin16 notes - it's probably worth a `removal` note as well, as it could potentially break existing apps where users have saved data in a non-XDG location.

cyphase commented:
I've been thinking for the past week that today might be the day I could get back to this and finish it up, but some work things intervened. It's still very much my intention to finish this ASAP.

I haven't taken the coin out of the sleeve, or looked at it since I got back. Not until I properly earn it. :laughing:

cyphase committed:
Combine comments.

cyphase committed:
Add test of directory creation.

Since Toga handles creating the root directories if they don't exist,
we don't want to create the parent folders here, so that it fails if
the directory wasn't created by Toga.

HalfWhitt commented:
@cyphase It looks like your local pre-commit probably needs updating. It's been long enough since you worked on this earlier that it's probably worth recreating your Python virtual environment, to make sure everything's up to date.
